File: binarysearchtree.py
class Node():

    def __init__(self, val=None):
        self.val = val
        self.left = None
        self.right = None

    def __iter__(self):
        return self.iterHelper(self)

    def iterHelper(self, root):
        if root:
            for val in self.iterHelper(root.left):
                yield val
            yield root.val
            for val in self.iterHelper(root.right):
                yield val

class BinarySearchTree():

    def __init__(self, name, root):
        self.root = root
        self.name = name

    def __iter__(self):
        return self.iterHelper(self.root)

    def iterHelper(self, root):
        if root:
            for val in self.iterHelper(root.left):
                yield val
            yield root.val
            for val in self.iterHelper(root.right):
                yield val

    def insert(self, val, root):
        if root.val != None :
            if val < root.val:
                if root.left:
                    self.insert(val,root.left)
                else:
                    root.left = Node(val)
            else:
                if root.right:
                    self.insert(val,root.right)
                else:
                    root.right = Node(val)
        else:
            root.val = val


    def add_all(self, *tup):
        for v in tup:
            self.insert(v, self.root)

    def __str__(self):
        ret = ""
        if self.root.val == None:
            return ret
        else:
            ret += "[" + self.name + "]"
            ret += self.strHelper(self.root)
            return ret

    def strHelper(self, root):
        res = ""
        if root != None:
            res += str(root.val)
            if root.left:
                res += " L:(" + self.strHelper(root.left)
                res += ")"
            if root.right:
                res += " R:(" + self.strHelper(root.right)
                res += ")"
        return res


class Merger():

    def __init__(self, name):
        self.name = name

    @classmethod
    def merge(cls, t1, t2):
        list = []
        l1 = iter(t1)
        l2 = iter(t2)
        e1 = next(l1)
        e2 = next(l2)
        while True:
            if e1 < e2:
                try:
                    list.append(e1)
                    e1 = next(l1)
                except StopIteration:
                    while True:
                        try:
                            list.append(e2)
                            e2 = next(l2)
                        except StopIteration:
                            break
                    break
            else:
                try:
                    list.append(e2)
                    e2 = next(l2)
                except StopIteration:
                    while True:
                        try:
                            list.append(e1)
                            e1 = next(l1)
                        except StopIteration:
                            break
                    break
        return list




if __name__ == "__main__":
    tree = BinarySearchTree(name="Oak", root=Node())
    tree.add_all(5, 3, 9, 0) # adds the elements in the order 5, 3, 9, and then 0
    print(tree)

if __name__ == "__main__":
    t1 = BinarySearchTree(name="Oak", root=Node())
    t2 = BinarySearchTree(name="Birch", root=Node())
    t1.add_all(5, 3, 9, 0)
    t2.add_all(1, 0, 10, 2, 7)
    for x in t1.root:
        print(x)
    for x in t2.root:
        print(x)
    print(Merger.merge(t1,t2))

File: BinarySearchTree.java
import java.util.*;


public class BinarySearchTree<T extends Comparable<T>> implements Iterable<T> {
    String name;
    Node<T> root;

    public BinarySearchTree(String name) {
        this.name = name;
    }
    public void addAll(List<T> lst) {
        for (T val : lst) {
            insert(val);
        }
    }
    public void insert(T val) {
        Node<T> node = new Node<T>(val);
        if (root == null) {
            root = node;
        }
        else{
            Node<T> curr = root;
            while(curr != null) {
                if (curr.getValue().compareTo(val) > 0) {
                    if (curr.getLeft() == null) {
                        curr.setLeft(node);
                        break;
                    } else {
                        curr = curr.getLeft();
                    }
                } else if(curr.getValue().compareTo(val) < 0) {
                    if (curr.getRight() == null) {
                        curr.setRight(node);
                        break;
                    } else {
                        curr = curr.getRight();
                    }
                }
            }
        }
    }

    class BstIterator implements Iterator<T> {
        Stack<Node<T>> stack;
        public BstIterator(Node<T> root) {
            stack = new Stack<Node<T>>();
            iterHelper(root);
        }
        private void iterHelper(Node<T> root){
            if(root!=null){
                if(root.getRight()!= null){
                    iterHelper(root.getRight());
                }
                stack.push(root);
                if(root.getLeft()!= null){
                    iterHelper(root.getLeft());
                }
            }
        }
        public boolean hasNext() {
            return stack.size() > 0;
        }
        public T next() {
            Node<T> ret = stack.pop();
            return ret.getValue();
        }
    }
    public Iterator<T> iterator() {
        return new BstIterator(root);
    }

    public String toString(){
        String ret = "";
        if(root==null){
            return ret;
        }
        else{
            ret = ret +  "[" + name + "]";
            ret = ret + strHelper(root);
        }
        return ret;
    }

    public String strHelper(Node<T> root){
        String res = "";
        if(root != null){

            res = res + root.getValue().toString();
            if(root.getLeft()!=null){
                res = res + " L:(" + strHelper(root.getLeft());
                res = res + ")";
            }
            if(root.getRight()!=null){
                res = res + " R:(" + strHelper(root.getRight());
                res = res + ")";
            }
        }
        return res;
    }
    public static <T extends Comparable<T>> List<T> merge(BinarySearchTree<T> t1, BinarySearchTree<T> t2){
        List<T> list = new ArrayList<T>();
        Thread thread = new Thread() {
            public void run() {
                t1.forEach(list::add);
            }
        };
        Thread thread2 = new Thread(){
            public void run(){
                t2.forEach(list::add);
            }
        };
        thread.start();
        thread2.start();
        try{
            thread.join();
            thread2.join();
            list.sort(null);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        return list;
    }

    public static void main(String... args) {
// each tree has a name, provided to its constructor
        BinarySearchTree<Integer> t1 = new BinarySearchTree<>("Oak");
// adds the elements to t1 in the order 5, 3, 0, and then 9
        t1.addAll(Arrays.asList(5, 3, 0, 9));
        BinarySearchTree<Integer> t2 = new BinarySearchTree<>("Maple");
// adds the elements to t2 in the order 9, 5, and then 10
        t2.addAll(Arrays.asList(9, 5, 10));
        System.out.println(t1); // see the expected output for exact format
        t1.forEach(System.out::println); // iteration in increasing order
        System.out.println(t2); // see the expected output for exact format
        t2.forEach(System.out::println); // iteration in increasing order
        BinarySearchTree<String> t3 = new BinarySearchTree<>("Cornucopia");
        t3.addAll(Arrays.asList("coconut", "apple", "banana", "plum", "durian",
                "no durians on this tree!", "tamarind"));
        System.out.println(t3); // see the expected output for exact format
        t3.forEach(System.out::println); // iteration in increasing order
    }
}
